core/src/test/java/org/elasticsearch/threadpool/ScheduleWithFixedDelayTests.java

```diff
@@ -0,0 +1,287 @@
+            assertTrue(counterMetExpected);
+
+            cancellable.cancel();
+            final int currentCount = counter.get();
+            assertTrue(cancellable.isCancelled());
+
+            // we cancel but there could be a currently running iteration, so we buffer by one value...
+            final boolean stillRunning = awaitBusy(() -> counter.get() > currentCount + 1, 250L, TimeUnit.MILLISECONDS);
+            assertFalse(stillRunning);
+        } finally {
+            threadPool.shutdown();
+        }
+    }
+
+    public void testThrowingRunnable() throws Exception {
+        final TimeValue delay = TimeValue.timeValueMillis(100L);
+        ThreadPool threadPool = mock(ThreadPool.class);
+        Runnable throwingRunnable = () ->  { throw new RuntimeException(); };
+        ReschedulingRunnable reschedulingRunnable = new ReschedulingRunnable(throwingRunnable, delay, Names.GENERIC, threadPool);
+        // nothing should be done with the threadpool on creation
+        verifyZeroInteractions(threadPool);
+
+        Cancellable cancellable = reschedulingRunnable.scheduleFirstRun();
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
+
+        // call run to mimic scheduler starting the runnable
+        RuntimeException runtimeException = expectThrows(RuntimeException.class, reschedulingRunnable::run);
+        assertEquals(null, runtimeException.getMessage());
+
+        // verify that there were no more interactions with the threadpool such as scheduling again
+        verifyNoMoreInteractions(threadPool);
+
+        assertTrue(cancellable.isCancelled());
+    }
+
+    public void testDoesNotRescheduleUntilExecutionFinished() throws Exception {
+        final TimeValue delay = TimeValue.timeValueMillis(100L);
+        final CountDownLatch startLatch = new CountDownLatch(1);
+        final CountDownLatch pauseLatch = new CountDownLatch(1);
+        ThreadPool threadPool = mock(ThreadPool.class);
+        final Runnable runnable = () ->  {
+            startLatch.countDown();
+            try {
+                pauseLatch.await();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        };
+        ReschedulingRunnable reschedulingRunnable = new ReschedulingRunnable(runnable, delay, Names.GENERIC, threadPool);
+        // nothing should be done with the threadpool on creation
+        verifyZeroInteractions(threadPool);
+
+        reschedulingRunnable.scheduleFirstRun();
+        Thread runThread = new Thread() {
+            @Override
+            public void run() {
+                reschedulingRunnable.run();
+            }
+        };
+        runThread.start();
+        startLatch.await();
+
+        // this call was made during scheduleFirstRun
+        verify(threadPool, times(1)).schedule(delay, Names.GENERIC, reschedulingRunnable);
```

Comment:
this test is really a validation of the behavior of the runnable in that it will not reschedule itself until after completion of the run method. I'll add some comments to make this a bit more clear